src/test/time/clock_nanosleep/test_clock_nanosleep.rs

```diff
@@ -0,0 +1,451 @@
+    });
+
+    std::thread::sleep(signaler_sleep_duration);
+    unsafe {
+        libc::syscall(libc::SYS_tkill, receiver.recv()?.as_raw(), libc::SIGUSR1);
+    }
+
+    sleeper.join().unwrap()
+}
+
+fn run_interrupted_sleeper(
+    clockid: libc::clockid_t,
+    flags: libc::c_int,
+    sig_dur: Duration,
+    sleep_dur: Duration,
+    sender: Sender<Pid>,
+) -> anyhow::Result<()> {
+    sender.send(nix::unistd::gettid()).unwrap();
+
+    let request = create_sleep_request(clockid, flags, sleep_dur)?;
+    let mut remain = libc::timespec {
+        tv_sec: 0,
+        tv_nsec: 0,
+    };
+
+    let before = SystemTime::now();
+    let rv = unsafe { libc::clock_nanosleep(clockid, flags, &request, &mut remain) };
+    let after = SystemTime::now();
+    ensure_ord!(libc::EINTR, ==, rv);
+
+    let tolerance = Duration::from_millis(30);
+
+    // Should have been interrupted after the signal duration.
+    let actual_dur = after.duration_since(before)?;
+    let interrupt_diff = duration_abs_diff(sig_dur, actual_dur);
+    ensure_ord!(interrupt_diff, <=, tolerance);
+
+    if flags != libc::TIMER_ABSTIME {
+        // Check reported time remaining.
+        let remain_expected = sleep_dur.checked_sub(sig_dur).unwrap();
+        let remain_actual = timespec_to_duration(remain);
+        let remain_diff = duration_abs_diff(remain_expected, remain_actual);
+        ensure_ord!(remain_diff, <=, tolerance);
+    }
+
+    Ok(())
+}
+
+fn install_signal_handler() -> anyhow::Result<()> {
+    unsafe {
+        nix::sys::signal::sigaction(
+            nix::sys::signal::SIGUSR1,
+            &nix::sys::signal::SigAction::new(
+                nop_sig_handler(),
+                nix::sys::signal::SaFlags::empty(),
+                nix::sys::signal::SigSet::empty(),
+            ),
+        )?
+    };
+    Ok(())
+}
+
+fn create_sleep_request(
+    clockid: libc::clockid_t,
+    flags: libc::c_int,
+    relative_dur: Duration,
+) -> anyhow::Result<libc::timespec> {
+    let absolute_offset = match flags == libc::TIMER_ABSTIME {
+        true => clock_now_duration(clockid)?,
+        false => Duration::ZERO,
+    };
+
+    let request_dur = absolute_offset.checked_add(relative_dur).unwrap();
+    let request = libc::timespec {
+        tv_sec: request_dur.as_secs().try_into()?,
+        tv_nsec: request_dur.subsec_nanos().into(),
+    };
+    Ok(request)
+}
+
+fn clock_now_duration(clockid: libc::clockid_t) -> anyhow::Result<Duration> {
+    let now = clock_now(clockid)?;
+    Ok(timespec_to_duration(now))
+}
+
+fn clock_now(clockid: libc::clockid_t) -> anyhow::Result<libc::timespec> {
+    let mut now = libc::timespec {
+        tv_sec: 0,
+        tv_nsec: 0,
+    };
+    let rv = unsafe { libc::clock_gettime(clockid, &mut now) };
+    ensure_ord!(rv, ==, 0);
+    Ok(now)
+}
+
+fn timespec_to_duration(ts: libc::timespec) -> Duration {
+    // valid tv_nsec values are [0, 999999999], which fit in a u32
+    Duration::new(ts.tv_sec as u64, ts.tv_nsec as u32)
```

Comment:
Optional:

* Using `Duration::from_secs` and `Duration::from_nanos` makes it clearer that the units are correct. 
* Using `try_into` ensures no surprise truncations etc

 `Duration::from_secs(ts.tv_sec.try_into().unwrap()) + Duration::from_nanos(ts.tv_nsec.try_into().unwrap())`

Reply:
Alternatively: `Duration::from(nix::sys::time::TimeSpec::from_timespec(ts))`